compile/compile_test.go

```diff
@@ -821,6 +821,201 @@ func TestCompilerPlanTargetUnmatchedEntrypoints(t *testing.T) {
+		wantEntrypoints map[string]struct{}
+	}{
+		{
+			note:        "rule annotation",
+			entrypoints: []string{},
+			modules: map[string]string{
+				"test.rego": `
+package test
+
+# METADATA
+# entrypoint: true
+p {
+	q[input.x]
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test/p": {},
+			},
+		},
+		{
+			note:        "package annotation",
+			entrypoints: []string{},
+			modules: map[string]string{
+				"test.rego": `
+# METADATA
+# entrypoint: true
+package test
+
+p {
+	q[input.x]
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test": {},
+			},
+		},
+		{
+			note:        "nested rule annotation",
+			entrypoints: []string{},
+			modules: map[string]string{
+				"test.rego": `
+package test
+
+import data.test.nested
+
+p {
+	q[input.x]
+	nested.p
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+				"test/nested.rego": `
+package test.nested
+
+# METADATA
+# entrypoint: true
+p {
+	q[input.x]
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test/nested/p": {},
+			},
+		},
+		{
+			note:        "nested package annotation",
+			entrypoints: []string{},
+			modules: map[string]string{
+				"test.rego": `
+package test
+
+import data.test.nested
+
+p {
+	q[input.x]
+	nested.p
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+				"test/nested.rego": `
+# METADATA
+# entrypoint: true
+package test.nested
+
+p {
+	q[input.x]
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test/nested": {},
+			},
+		},
+		{
+			note:        "mixed manual entrypoints + annotation entrypoints",
+			entrypoints: []string{"test/p"},
+			modules: map[string]string{
+				"test.rego": `
+package test
+
+import data.test.nested
+
+p {
+	q[input.x]
+	nested.p
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+				"test/nested.rego": `
+# METADATA
+# entrypoint: true
+package test.nested
+
+p {
+	q[input.x]
+}
+
+q[1]
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test/nested": {},
+				"test/p":      {},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+			test.WithTempFS(tc.modules, func(root string) {
+
+				compiler := New().
+					WithPaths(root).
+					WithTarget("plan").
+					WithEntrypoints(tc.entrypoints...).
+					WithRegoAnnotationEntrypoints(true).
+					WithPruneUnused(true)
+				err := compiler.Build(context.Background())
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				// Ensure we have the right number of entrypoints.
+				if len(compiler.entrypoints) != len(tc.wantEntrypoints) {
+					t.Fatalf("Wrong number of entrypoints. Expected %v, got %v.", tc.wantEntrypoints, compiler.entrypoints)
+				}
+
+				// Ensure those entrypoints match the ones we expect.
+				for _, entrypoint := range compiler.entrypoints {
+					if _, found := tc.wantEntrypoints[entrypoint]; !found {
+						t.Fatalf("Unexpected entrypoint '%s'", entrypoint)
+					}
+				}
+			})
+		})
+	}
+}
```

Comment:
Let's have a ref head rule, too! 😃

Reply:
This was a great suggestion, because as I had feared, I wasn't using `ref.Head.Ref()` to extract the refs for entrypoints, and was instead relying on the older `ref.Head.Name` style that can fail on newer-style ref head rules.

Reply:
Rolled in 2x new tests for ref heads. @srenatus if you can think of more clever cases I should try, please LMK, and I'll roll them in to another round of changes. :smile: